src/libmongoc/tests/unified/runner.c

```diff
@@ -20,89 +20,729 @@
 #include "test-libmongoc.h"
 #include <signal.h>
 
-/* The test context is a global storing state for each test.
+/* TODO: put in separate files? runner.c, file.c, test.c, diagnostics.c */
+
+/* test_runner_t, test_file_t, and test_t model the types described in the "Test
+ * Runner Implementation" section of the Unified Test Format specification. */
+typedef struct {
+   mongoc_client_t *internal_client;
+   semver_t server_version;
+   /* topology_type may be "single", "replicaset", "sharded", or
+    * "sharded-replicaset". */
+   const char *topology_type;
+   mongoc_array_t server_ids;
+} test_runner_t;
+
+typedef struct {
+   test_runner_t *test_runner;
+
+   const char *description;
+   semver_t schema_version;
+   bson_t *run_on_requirements;
+   bson_t *create_entities;
+   bson_t *initial_data;
+   bson_t *tests;
+} test_file_t;
+
+typedef struct {
+   test_file_t *test_file;
+
+   const char *description;
+   bson_t *run_on_requirements;
+   const char *skip_reason;
+   bson_t *operations;
+   bson_t *expect_events;
+   bson_t *outcome;
+} test_t;
+
+/* test_diagnostics is a global storing current test state.
  * When an assertion fails, causing an abort signal, the test
- * context can be logged. */
+ * diagnostics are logged. */
 struct {
-   bson_t *test;
-   const char *description;
-   bson_t command_started_events;
-   bson_t command_succeeded_events;
-   bson_t command_failed_events;
-} test_ctx;
+   test_runner_t *test_runner;
+   test_file_t *test_file;
+   test_t *test;
+} test_diagnostics;
```

Comment:
It's not yet clear to me whether this is useful. The spec recommends [not aborting while a test is running so server state can be cleaned up](https://github.com/mongodb/specifications/blob/master/source/unified-test-format/unified-test-format.rst#executing-a-test):

> If at any point while executing this test an unexpected error is encountered or an assertion fails, the test runner MUST consider this test to have failed and SHOULD continue with the instructions in this section to ensure that the test environment is cleaned up (e.g. disable fail points, kill sessions) while also forgoing any additional assertions.

I plan to have`test_run` return an error instead of aborting when something within the test fails. So, if that means the signal handler diagnostics is not useful, I may remove it later.